tokio-timer/src/timer/mod.rs

```diff
@@ -0,0 +1,645 @@
+    ///
+    /// The `Handle` is how `Sleep` instances are created. The `Sleep` instances
+    /// can either be created directly or the `Handle` instance can be passed to
+    /// `with_default`, setting the timer as the default timer for the execution
+    /// context.
+    pub fn handle(&self) -> Handle {
+        Handle::new(Arc::downgrade(&self.inner))
+    }
+
+    /// Performs one iteration of the timer loop.
+    ///
+    /// This function must be called repeatedly in order for the `Timer`
+    /// instance to make progress. This is where the work happens.
+    ///
+    /// The `Timer` will use the `Park` instance that was specified in [`new`]
+    /// to block the current thread until the next `Sleep` instance elapses. One
+    /// call to `turn` results in at most one call to `park.park()`.
+    ///
+    /// # Return
+    ///
+    /// On success, `Ok(Turn)` is returned, where `Turn` is a placeholder type
+    /// that currently does nothing but may, in the future, have functions add
+    /// to provide information about the call to `turn`.
+    ///
+    /// If the call to `park.park()` fails, then `Err` is returned with the
+    /// error.
+    ///
+    /// [`new`]: #method.new
+    pub fn turn(&mut self, max_wait: Option<Duration>) -> Result<Turn, T::Error> {
+        match max_wait {
+            Some(timeout) => self.park_timeout(timeout)?,
+            None => self.park()?,
+        }
+
+        Ok(Turn(()))
+    }
+
+    /// Returns the instant at which the next timeout expires.
+    fn next_expiration(&self) -> Option<Expiration> {
+        // Check all levels
+        for level in 0..NUM_LEVELS {
+            if let Some(expiration) = self.levels[level].next_expiration(self.elapsed) {
+                // There cannot be any expirations at a higher level that happen
+                // before this one.
+                debug_assert!({
+                    let mut res = true;
+
+                    for l2 in (level+1)..NUM_LEVELS {
+                        if let Some(e2) = self.levels[l2].next_expiration(self.elapsed) {
+                            if e2.deadline < expiration.deadline {
+                                res = false;
+                            }
+                        }
+                    }
+
+                    res
+                });
+
+                return Some(expiration);
+            }
+        }
+
+        None
+    }
+
+    /// Converts an `Expiration` to an `Instant`.
+    fn expiration_instant(&self, expiration: &Expiration) -> Instant {
+        self.inner.start + Duration::from_millis(expiration.deadline)
+    }
+
+    /// Run timer related logic
+    fn process(&mut self) {
+        let now = ms(self.now.now() - self.inner.start, Round::Down);
+
+        loop {
+            let expiration = match self.next_expiration() {
+                Some(expiration) => expiration,
+                None => break,
+            };
+
+            if expiration.deadline > now {
```

Comment:
TIOLI, but this could be collapsed into the match, as in
```rust
let expiration = match self.next_expiration() {
    Some(expiration) if expiration.deadline <= now => expiration,
    _ => break,
};
```
To me, consolidating the conditions under which we can break seems clearer, but TIOLI.

Reply:
I'm usually not a fan of `if` clauses in match blocks. It makes it harder for me to unpack when reading.

Also... `<= now =>` hands out operator?

Reply:
> I'm usually not a fan of if clauses in match blocks. It makes it harder for me to unpack when reading.

Fair, I guess this is just a matter of personal taste then.

> Also... `<= now =>` hands out operator?

👐